File: BAEKJOON/13459 - 구슬 탈출/구슬 탈출.py
from collections import deque  # BFS를 위한 deque 라이브러리 임포트

dx = [-1, 1, 0, 0]  # 상, 하, 좌, 우 이동 방향 벡터 (x축)
dy = [0, 0, -1, 1]  # 상, 하, 좌, 우 이동 방향 벡터 (y축)


# 구슬을 기울여서 이동시키는 함수
def move(x, y, dx, dy, board):
    count = 0  # 이동 거리(구슬이 몇 칸 이동했는지)
    # 구슬이 벽(#)을 만나거나 구멍(O)에 빠질 때까지 이동
    while board[x + dx][y + dy] != "#" and board[x][y] != "O":
        x += dx  # x 방향으로 이동
        y += dy  # y 방향으로 이동
        count += 1  # 이동 횟수 증가
    return x, y, count  # 이동한 위치와 이동 거리 반환


# BFS(너비 우선 탐색)로 구슬 탈출 시뮬레이션을 수행하는 함수
def bfs(board, rx, ry, bx, by):
    queue = deque()  # BFS를 위한 큐 초기화
    queue.append(
        (rx, ry, bx, by, 0)
    )  # 초기 상태(빨간 구슬, 파란 구슬 위치, 횟수)를 큐에 추가
    visited = set()  # 방문 상태를 저장하는 집합
    visited.add((rx, ry, bx, by))  # 초기 상태를 방문 상태에 추가

    while queue:
        rx, ry, bx, by, depth = queue.popleft()  # 큐에서 상태를 하나 꺼냄

        if depth >= 10:  # 10번 이상 기울이면 실패로 간주
            return 0

        for i in range(4):  # 네 방향(상, 하, 좌, 우)으로 기울이기
            nrx, nry, rcnt = move(rx, ry, dx[i], dy[i], board)  # 빨간 구슬 이동
            nbx, nby, bcnt = move(bx, by, dx[i], dy[i], board)  # 파란 구슬 이동

            if board[nbx][nby] == "O":  # 파란 구슬이 구멍에 빠지면 실패 (계속 진행)
                continue
            if board[nrx][nry] == "O":  # 빨간 구슬이 구멍에 빠지면 성공
                return 1

            # 두 구슬이 같은 위치에 있을 경우
            if nrx == nbx and nry == nby:
                # 더 많이 이동한 구슬을 한 칸 뒤로 이동
                if rcnt > bcnt:
                    nrx -= dx[i]
                    nry -= dy[i]
                else:
                    nbx -= dx[i]
                    nby -= dy[i]

            # 현재 상태가 방문하지 않은 상태라면 큐에 추가
            if (nrx, nry, nbx, nby) not in visited:
                visited.add((nrx, nry, nbx, nby))  # 방문 상태 기록
                queue.append((nrx, nry, nbx, nby, depth + 1))  # 다음 상태를 큐에 추가

    return 0  # 10번 이내에 성공하지 못한 경우 실패


# 문제를 해결하는 함수 (입력 데이터를 기반으로 보드 상태 초기화 및 BFS 호출)
def solve(board):
    N = len(board)  # 보드의 세로 크기
    M = len(board[0])  # 보드의 가로 크기
    rx = ry = bx = by = 0  # 빨간 구슬, 파란 구슬 위치 초기화

    for i in range(N):  # 보드의 각 행을 순회
        for j in range(M):  # 보드의 각 열을 순회
            if board[i][j] == "R":  # 빨간 구슬 위치 찾기
                rx, ry = i, j
            if board[i][j] == "B":  # 파란 구슬 위치 찾기
                bx, by = i, j

    return bfs(board, rx, ry, bx, by)  # 초기 구슬 위치를 바탕으로 BFS 수행 및 결과 반환


# 프로그램의 진입점
if __name__ == "__main__":
    N, M = map(int, input().split())  # 보드의 세로(N), 가로(M) 크기 입력
    board = [list(input().strip()) for _ in range(N)]  # N줄에 걸쳐 보드 상태 입력
    print(solve(board))  # 구슬 탈출 가능 여부 출력

File: BAEKJOON/13459 - 구슬 탈출/구슬 탈출.cs
using System;
using System.Collections.Generic;

class MarbleEscape {
    static int[] dx = { -1, 1, 0, 0 }; // 상, 하, 좌, 우 방향 벡터
    static int[] dy = { 0, 0, -1, 1 };

    // 구슬 이동 함수
    static (int, int, int) Move(int x, int y, int dx, int dy, char[,] board) {
        int count = 0; // 이동 거리
        while (board[x + dx, y + dy] != '#' && board[x, y] != 'O') { // 벽이나 구멍을 만나기 전까지 이동
            x += dx;
            y += dy;
            count++;
        }
        return (x, y, count); // 최종 위치와 이동 거리 반환
    }

    // BFS로 상태 탐색
    static int Bfs(char[,] board, int rx, int ry, int bx, int by) {
        Queue<(int, int, int, int, int)> queue = new(); // BFS 상태 저장 큐
        HashSet<(int, int, int, int)> visited = new();  // 방문한 상태 저장 집합
        queue.Enqueue((rx, ry, bx, by, 0)); // 초기 상태 추가
        visited.Add((rx, ry, bx, by));      // 초기 상태 방문 기록

        while (queue.Count > 0) {
            var (crx, cry, cbx, cby, depth) = queue.Dequeue(); // 큐에서 상태 꺼냄

            if (depth >= 10) return 0; // 10번 이상 기울인 경우 실패로 간주

            for (int i = 0; i < 4; i++) { // 네 방향으로 기울이기
                var (nrx, nry, rcnt) = Move(crx, cry, dx[i], dy[i], board); // 빨간 구슬 이동
                var (nbx, nby, bcnt) = Move(cbx, cby, dx[i], dy[i], board); // 파란 구슬 이동

                if (board[nbx, nby] == 'O') continue; // 파란 구슬이 구멍에 빠지면 실패
                if (board[nrx, nry] == 'O') return 1; // 빨간 구슬이 구멍에 빠지면 성공

                // 두 구슬이 같은 위치에 도달한 경우
                if (nrx == nbx && nry == nby) {
                    if (rcnt > bcnt) { // 빨간 구슬이 더 멀리 이동한 경우 뒤로 이동
                        nrx -= dx[i];
                        nry -= dy[i];
                    } else { // 파란 구슬이 더 멀리 이동한 경우 뒤로 이동
                        nbx -= dx[i];
                        nby -= dy[i];
                    }
                }

                // 방문하지 않은 상태만 큐에 추가
                if (!visited.Contains((nrx, nry, nbx, nby))) {
                    visited.Add((nrx, nry, nbx, nby)); // 새로운 상태 방문 기록
                    queue.Enqueue((nrx, nry, nbx, nby, depth + 1)); // 새로운 상태 큐에 추가
                }
            }
        }

        return 0; // 10번 이내에 성공하지 못한 경우 실패
    }

    // 문제 해결 함수
    static int Solve(char[,] board, int n, int m) {
        int rx = 0, ry = 0, bx = 0, by = 0;

        // 보드에서 빨간 구슬과 파란 구슬의 초기 위치 찾기
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                if (board[i, j] == 'R') {
                    rx = i; ry = j;
                }
                if (board[i, j] == 'B') {
                    bx = i; by = j;
                }
            }
        }

        return Bfs(board, rx, ry, bx, by); // BFS를 통해 문제 해결
    }

    // 메인 함수
    static void Main() {
        var input = Console.ReadLine().Split(); // 보드 크기 입력
        int n = int.Parse(input[0]);
        int m = int.Parse(input[1]);

        char[,] board = new char[n, m]; // 2D 보드 배열 선언

        for (int i = 0; i < n; i++) { // 보드 상태 입력
            var line = Console.ReadLine();
            for (int j = 0; j < m; j++) {
                board[i, j] = line[j];
            }
        }

        Console.WriteLine(Solve(board, n, m)); // 결과 출력
    }
}
